sql-plugin/src/main/scala/com/nvidia/spark/rapids/NvcompLZ4CompressionCodec.scala

```diff
@@ -0,0 +1,155 @@
+      var compressedSize: Long = 0L
+      val outputSize = LZ4Compressor.getOutputSize(input, CompressionType.CHAR, LZ4_CHUNK_SIZE,
+        tempBuffer)
+      closeOnExcept(DeviceMemoryBuffer.allocate(outputSize)) { outputBuffer =>
+        compressedSize = LZ4Compressor.compress(input, CompressionType.CHAR, LZ4_CHUNK_SIZE,
+          tempBuffer, outputBuffer, Cuda.DEFAULT_STREAM)
+        (compressedSize, outputBuffer)
+      }
+    }
+  }
+
+  /**
+   * Decompress data that was compressed with nvcomp's LZ4-GPU codec
+   * @param outputBuffer where the uncompressed data will be written
+   * @param inputBuffer buffer of compressed data to decompress
+   */
+  def decompress(outputBuffer: DeviceMemoryBuffer, inputBuffer: DeviceMemoryBuffer): Unit = {
+    withResource(Decompressor.getMetadata(inputBuffer, Cuda.DEFAULT_STREAM)) { metadata =>
+      val outputSize = Decompressor.getOutputSize(metadata)
+      if (outputSize != outputBuffer.getLength) {
+        throw new IllegalStateException(
+          s"metadata uncompressed size is $outputSize, buffer size is ${outputBuffer.getLength}")
+      }
+      val tempSize = Decompressor.getTempSize(metadata)
+      withResource(DeviceMemoryBuffer.allocate(tempSize)) { tempBuffer =>
+        Decompressor.decompressAsync(inputBuffer, tempBuffer, metadata, outputBuffer,
+          Cuda.DEFAULT_STREAM)
+      }
+    }
+  }
+}
+
+class BatchedNvcompLZ4Compressor(maxBatchMemorySize: Long)
+    extends BatchedTableCompressor(maxBatchMemorySize) {
+  override protected def compress(tables: Array[ContiguousTable]): Array[CompressedTable] = {
+    val inputBuffers: Array[BaseDeviceMemoryBuffer] = tables.map(_.getBuffer)
+    val compressionResult = BatchedLZ4Compressor.compress(inputBuffers,
+      NvcompLZ4CompressionCodec.LZ4_CHUNK_SIZE, Cuda.DEFAULT_STREAM)
+    val compressedTables = try {
+      val buffers = compressionResult.getCompressedBuffers
+      val compressedSizes = compressionResult.getCompressedSizes
+      buffers.zipWithIndex.map { case (buffer, i) =>
+        val contigTable = tables(i)
+        val compressedSize = compressedSizes(i)
+        val meta = MetaUtils.buildTableMeta(
+          tableId = 0,
+          contigTable.getTable,
+          contigTable.getBuffer,
+          CodecType.NVCOMP_LZ4,
+          compressedSize)
+        CompressedTable(compressedSize, meta, buffer)
+      }
+    } catch {
+      case t: Throwable =>
+        compressionResult.getCompressedBuffers.safeClose()
+        throw t
+    }
+
+    // output buffer sizes were estimated and probably significantly oversized, so copy any
+    // oversized buffers to properly sized buffers in order to release the excess memory.
+    resizeOversizedOutputs(compressedTables)
+  }
+}
+
+class BatchedNvcompLZ4Decompressor(maxBatchMemory: Long)
+    extends BatchedBufferDecompressor(maxBatchMemory) {
+  override val codecId: Byte = CodecType.NVCOMP_LZ4
+
+  // TODO: Need to pass stream as arg, make name async
```

Comment:
do we want to file followup issue to track this?

Reply:
This is something I should address here, will fix.

Reply:
I added a stream parameter to the table compression APIs and renamed the decompression methods to include `Async` in the name.